### PR TITLE
Implement Helpik service-component

### PR DESCRIPTION
**Pull request by Themanwhosmellslikesugar**

### CODE REVIEW

File: src/api/http/helpik.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import axios from "axios";
+
+const synopsis = (page) => (
```
Comment (karpa4o4) [inline src/api/http/helpik.js]: Let't rename function to `fetchSynopsis` and `page` argument to `pageName`.

File: webpack.config.babel.js [hunk 2]
```diff
@@ -70,12 +71,18 @@ module.exports = {
...
+            MEDIA_WIKI_URL: JSON.stringify(MEDIA_WIKI_URL),
+        }),
     ],
     devServer: {
         host: HOST,
         contentBase: path.resolve(__dirname, "dist"),
         publicPath: "/",
         historyApiFallback: true,
+        headers: {
+            "Access-Control-Allow-Origin": "http://localhost:8010/",
```
Comment (karpa4o4) [inline webpack.config.babel.js]: Make `Access-Control-Allow-Origin` value dynamically substituting from `MEDIA_WIKI_URL`.

File: webpack.config.babel.js [hunk 3]
```diff
@@ -7,6 +7,7 @@ const srcPath = path.resolve(rootPath, "./src");
 
 const HOST = process.env.HOST || "localhost";
 const { BM_RPC_ADDR } = process.env;
+const { MEDIA_WIKI_URL } = process.env;
```
Comment (karpa4o4) [inline webpack.config.babel.js]: You can combine this line with the previous one.

File: src/common/services/Helpik/Helpik.jsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { faQuestionCircle } from "@fortawesome/free-solid-svg-icons";
+
+import Popover from "react-bootstrap/Popover";
+import Overlay from "react-bootstrap/Overlay";
+
+import OutsideAlerter from "common/components/OutsideAlerter";
+
+import synopsis from "api/http/helpik";
```
Comment (karpa4o4) [inline src/common/services/Helpik/Helpik.jsx]: Import objects from "api/http/helpik" as `synopsis as API`.

File: src/common/components/OutsideAlerter.jsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    componentDidMount() {
+        document.addEventListener("mousedown", this.handleClickOutside);
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener("mousedown", this.handleClickOutside);
+    }
+
+    setWrapperRef(node) {
```
Comment (karpa4o4) [inline src/common/components/OutsideAlerter.jsx]: This is redunant function.

File: src/common/components/OutsideAlerter.jsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import React, { Component } from "react";
```
Comment (karpa4o4) [inline src/common/components/OutsideAlerter.jsx]: Let's move this component to `src/common/services/Helpik/components/`

File: src/common/components/OutsideAlerter.jsx [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+export default class OutsideAlerter extends Component {
+    constructor(props) {
+        super(props);
+
+        this.setWrapperRef = this.setWrapperRef.bind(this);
+        this.handleClickOutside = this.handleClickOutside.bind(this);
+    }
+
+    componentDidMount() {
+        document.addEventListener("mousedown", this.handleClickOutside);
```
Comment (karpa4o4) [inline src/common/components/OutsideAlerter.jsx]: Whe the `mousedown` event is being listen, and not the `click`?

File: webpack.config.babel.js (hunk 2 above)
Comment (eugulixes) [reply webpack.config.babel.js]: `MEDIAWIKI_URL`

File: src/api/http/helpik.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import axios from "axios";
+
+export const fetchSynopsis = (pageName, languageName) => (
+    /* global HELPIK_URL */
+    axios.get(`${HELPIK_URL}?pageName=${pageName}&language=${languageName}`)
```
Comment (karpa4o4) [inline src/api/http/helpik.js]: Let's pass params like here:
```
axios.get('<url>', {
    params: {},
})
```

File: src/api/http/helpik.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import axios from "axios";
+
+export const fetchSynopsis = (pageName, languageName) => (
```
Comment (karpa4o4) [inline src/api/http/helpik.js]: The name `fetchHelpikData` is more appropriate.

File: src/api/http/helpik.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+    /* global HELPIK_URL */
+    axios.get(`${HELPIK_URL}?pageName=${pageName}&language=${languageName}`)
+        .then((response) => {
```
Comment (karpa4o4) [inline src/api/http/helpik.js]: This funtion is redunant, because you use `then` when calling `fetchSynopsis`.

File: webpack.config.babel.js [hunk 11]
```diff
@@ -70,12 +70,18 @@ module.exports = {
         new webpack.DefinePlugin({
             BM_RPC_ADDR: JSON.stringify(BM_RPC_ADDR),
         }),
+        new webpack.DefinePlugin({
```
Comment (karpa4o4) [inline webpack.config.babel.js]: You don't need to create another one instance of `webpack.DefinePlugin`. You can define `HELPIK_URL` in the instance above.

File: src/api/http/helpik.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import axios from "axios";
+import { helpikURL } from "config/main";
+
+export const fetchHelpikData = (pageName, languageName) => {
+    const response = axios.get(`${helpikURL}/get_synopsis/`, {
```
Comment (karpa4o4) [inline src/api/http/helpik.js:5]: Please, make the request asynchronously.

File: src/common/services/Helpik/Helpik.jsx [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    handleIconClick(event) {
+        const { helpikStatus } = this.state;
+        const currentTarget = event.target;
+        this.setState(() => ({
+            helpikStatus: !helpikStatus,
+            target: currentTarget,
+        }));
+    }
+
+    OutsideAlerterCallback() {
```
Comment (karpa4o4) [inline src/common/services/Helpik/Helpik.jsx:56]: You need to start the function name with a small letter

File: src/common/services/Helpik/Helpik.jsx [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        this.OutsideAlerterCallback = this.OutsideAlerterCallback.bind(this);
+    }
+
+    componentDidMount() {
+        const { pageName } = this.props;
+        API.fetchHelpikData(pageName, "en")
+            .then((response) => {
+                this.setState(() => ({
+                    helpikContent: response.data.text,
+                    helpikURL: response.data.url.replace("api.php", ""),
```
Comment (karpa4o4) [inline src/common/services/Helpik/Helpik.jsx:36]: Where does the `api.php` come from here? Can we not send it in the response from the server?

File: src/common/services/Helpik/Helpik.jsx [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+                        </Popover>
+                    </Overlay>
+                </span>
+            </span>
+        );
+    }
+}
+
+Helpik.propTypes = {
+    id: PropTypes.string,
```
Comment (karpa4o4) [inline src/common/services/Helpik/Helpik.jsx:107]: How is this parameter used?

File: src/common/services/Helpik/Helpik.jsx [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+import OutsideAlerter from "./components/OutsideAlerter";
+
+export default class Helpik extends Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            helpikStatus: false,
+            helpikContent: "",
+            helpikURL: "",
+            helpikVisability: "d-none",
```
Comment (karpa4o4) [inline src/common/services/Helpik/Helpik.jsx:21]: Let's make it a Boolean value.

File: src/common/services/Helpik/Helpik.jsx [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+        this.setState(() => ({
+            helpikStatus: false,
+        }));
+    }
+
+    render() {
+        const {
+            helpikStatus, helpikContent, helpikURL, helpikVisability, target,
+        } = this.state;
+        const iconColor = !helpikStatus ? "text-gray" : "text-black";
```
Comment (karpa4o4) [inline src/common/services/Helpik/Helpik.jsx:61]: Logical negation is unnecessary here.

File: src/common/services/Helpik/styles.scss [hunk 18]
```diff
@@ -0,0 +1,13 @@
+.helpik-main {
+  @media (hover: hover) and (pointer: fine) {
```
Comment (karpa4o4) [inline src/common/services/Helpik/styles.scss:2]: Elegant solution!

File: src/common/services/Helpik/components/OutsideAlerter.jsx [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    render() {
+        const { children } = this.props;
+        return <span ref={this.wrapperRef}>{children}</span>;
+    }
+}
+
+OutsideAlerter.propTypes = {
+    children: PropTypes.element.isRequired,
+    outsideAlerterCallback: PropTypes.func.isRequired,
+    iconTarget: PropTypes.objectOf(PropTypes.object).isRequired,
```
Comment (karpa4o4) [inline src/common/services/Helpik/components/OutsideAlerter.jsx:37]: Will `PropTypes.object` not be enough?

File: src/common/services/Helpik/styles.scss [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+    .icon-hover:hover {
+      color: black !important;
+      cursor: pointer;
+    }
+  }
+}
```
Comment (karpa4o4) [inline src/common/services/Helpik/styles.scss:8]: Please, add empty line after this.

File: src/common/services/Helpik/styles.scss [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+.helpik-popover {
```
Comment (karpa4o4) [inline src/common/services/Helpik/styles.scss:9]: Add a comment describing the problem that this code solves.

File: src/common/services/Helpik/components/OutsideAlerter.jsx (hunk 19 above)
Comment (Themanwhosmellslikesugar) [reply src/common/services/Helpik/components/OutsideAlerter.jsx:37]: eslint requires type checking inside an object, else will be an error `Prop type `object` is forbidden`

Comment (karpa4o4) [reply src/common/services/Helpik/components/OutsideAlerter.jsx:37]: :+1:

File: src/common/services/Helpik/Helpik.jsx [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+            .then((response) => {
+                this.setState(() => ({
+                    helpikContent: response.data.text,
+                    helpikURL: response.data.url,
+                    helpikVisability: true,
+                }));
+            });
+    }
+
+    setHelpikRef(helpik) {
```
Comment (karpa4o4) [inline src/common/services/Helpik/Helpik.jsx:42]: This function is unnecessary.

File: src/common/services/Helpik/styles.scss [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+    .icon-hover:hover {
+      color: black !important;
+      cursor: pointer;
+    }
+  }
+}
+
+// This is bad solution to use margin for
+// Popper library.
+// Solution is use prop 'popperConfig' in Ovelay, but
```
Comment (eugulixes) [inline src/common/services/Helpik/styles.scss:12]: `Solution is to use...`

**Pull request merged**